base/src/test/java/org/aya/repr/ShapeMatcherTest.java

```diff
@@ -52,6 +52,17 @@ data List (A : Type)
       """);
   }
 
+  @Test
+  public void matchWeirdList() {
+    match(true, AyaShape.AyaListShape.DATA_LIST, "data List {A : Type} | nil | cons A (List {A})");
+    match(true, AyaShape.AyaListShape.DATA_LIST, "data List (A : Type) | nil | cons {A} (List A)");
+    match(true, AyaShape.AyaListShape.DATA_LIST, "data List (A : Type) | nil | cons A {List A}");
+    match(true, AyaShape.AyaListShape.DATA_LIST, "data List {A : Type} | nil | cons {A} (List {A})");
+    match(true, AyaShape.AyaListShape.DATA_LIST, "data List {A : Type} | nil | cons A {List {A}}");
+    match(true, AyaShape.AyaListShape.DATA_LIST, "data List (A : Type) | nil | cons {A} {List A}");
+    match(true, AyaShape.AyaListShape.DATA_LIST, "data List {A : Type} | nil | cons {A} {List {A}}");
+  }
```

Comment:
Should we support them? The code shapes are designed to optimize common data structures, while these lists are not common.

Reply:
I don't know, ask @ice1000

Reply:
I do think that ctors like `| suc {Nat}` look weird and useless, but we need to support `List {A}`

Reply:
I agree that `suc {Nat}` looks weird but it becomes less weird when the type becomes interesting (for example, the type of "less than" relation is isomorphic to Nat)

Reply:
Agreed!